### PR TITLE
[CELEBORN-804] ShuffleClient should cleanup shuffle infos when trigger unregisterShuffle.

### PR DESCRIPTION
**Pull request by AngersZhuuuu**
### What changes were proposed in this pull request?

After discussion, we make sure that `shuffleManager.unregisterShuffle()` will be triggered by Spark both in driver and executor. In this pr:

  1. Add shuffle client both in driver and executor side in ShuffleManager
  2. ShuffleClient call cleanupShuffle() when trigger `unregisterShuffle`.

This replaced https://github.com/apache/incubator-celeborn/pull/1719

### CODE REVIEW

**Comment by AngersZhuuuu** [comment]
ping @waitinfuture @FMX @RexXiong @pan3793 @cfmcgrady

File: client/src/main/scala/org/apache/celeborn/client/LifecycleManager.scala [hunk 1]
```diff
@@ -198,11 +198,6 @@ class LifecycleManager(val appUniqueId: String, val conf: CelebornConf) extends
     case StageEnd(shuffleId) =>
       logInfo(s"Received StageEnd request, shuffleId $shuffleId.")
       handleStageEnd(shuffleId)
-    case pb: PbUnregisterShuffle =>
```
Comment (cfmcgrady) [inline client/src/main/scala/org/apache/celeborn/client/LifecycleManager.scala:201]: shall we also update the `proto` file? or add a comment to the `proto` file for future cleanup?

File: client/src/main/java/org/apache/celeborn/client/ShuffleClient.java [hunk 2]
```diff
@@ -150,7 +152,7 @@ public abstract CelebornInputStream readPartition(
   public abstract CelebornInputStream readPartition(
       int shuffleId, int partitionId, int attemptNumber) throws IOException;
 
-  public abstract boolean unregisterShuffle(int shuffleId, boolean isDriver);
+  public abstract boolean unregisterShuffle(int shuffleId);
```
Comment (RexXiong) [inline client/src/main/java/org/apache/celeborn/client/ShuffleClient.java:155]: Little confuse for the method(unregisterShuffle) now, better to change it.

File: client-spark/spark-2/src/main/java/org/apache/spark/shuffle/celeborn/SparkShuffleManager.java [hunk 3]
```diff
@@ -150,10 +141,12 @@ public boolean unregisterShuffle(int shuffleId) {
     if (appUniqueId == null) {
       return true;
     }
-    if (shuffleClient == null) {
+    if (!isDriver()) {
```
Comment (waitinfuture) [inline client-spark/spark-2/src/main/java/org/apache/spark/shuffle/celeborn/SparkShuffleManager.java:144]: When will this happen, for local mode?

File: client/src/main/scala/org/apache/celeborn/client/LifecycleManager.scala (hunk 1 above)
Comment (waitinfuture) [reply client/src/main/scala/org/apache/celeborn/client/LifecycleManager.scala:201]: `PbUnregisterShuffle` is also used in Master, so I guess we can't delete the message.

File: client-spark/spark-2/src/main/java/org/apache/spark/shuffle/celeborn/SparkShuffleManager.java [hunk 4]
```diff
@@ -163,8 +156,8 @@ public ShuffleBlockResolver shuffleBlockResolver() {
 
   @Override
   public void stop() {
-    if (shuffleClient != null) {
-      shuffleClient.shutdown();
+    if (isDriver()) {
```
Comment (waitinfuture) [inline client-spark/spark-2/src/main/java/org/apache/spark/shuffle/celeborn/SparkShuffleManager.java:159]: When will `isDriver==false` happen? And if it is true, why do we need to reset ShuffleClient?

File: client/src/main/java/org/apache/celeborn/client/ShuffleClient.java (hunk 2 above)
Comment (waitinfuture) [reply client/src/main/java/org/apache/celeborn/client/ShuffleClient.java:155]: Maybe cleanupShuffle

Comment (AngersZhuuuu) [reply client/src/main/java/org/apache/celeborn/client/ShuffleClient.java:155]: Done

File: client-spark/spark-2/src/main/java/org/apache/spark/shuffle/celeborn/SparkShuffleManager.java (hunk 3 above)
Comment (AngersZhuuuu) [reply client-spark/spark-2/src/main/java/org/apache/spark/shuffle/celeborn/SparkShuffleManager.java:144]: Not sure, it's just follow old logic.

File: client-spark/spark-2/src/main/java/org/apache/spark/shuffle/celeborn/SparkShuffleManager.java (hunk 4 above)
Comment (AngersZhuuuu) [reply client-spark/spark-2/src/main/java/org/apache/spark/shuffle/celeborn/SparkShuffleManager.java:159]: Reset always for unit test, follow old logic

File: client-spark/spark-3/src/main/java/org/apache/spark/shuffle/celeborn/SparkShuffleManager.java [hunk 5]
```diff
@@ -154,13 +153,15 @@ public boolean unregisterShuffle(int shuffleId) {
...
     }
-    if (appUniqueId == null) {
-      return true;
+    // For Spark driver side trigger unregister shuffle.
+    if (lifecycleManager != null) {
+      lifecycleManager.unregisterShuffle(shuffleId);
     }
-    if (shuffleClient == null) {
-      return false;
+    // For Spark executor side cleanup shuffle related info.
```
Comment (pan3793) [inline client-spark/spark-3/src/main/java/org/apache/spark/shuffle/celeborn/SparkShuffleManager.java:160]: this is the key change, it's much simpler than the previous approach, I am wondering if we can revert the previous commit and re-submit this PR to make the change clearer.

File: client-spark/spark-3/src/main/java/org/apache/spark/shuffle/celeborn/SparkShuffleManager.java [hunk 6]
```diff
@@ -116,8 +116,7 @@ private void initializeLifecycleManager() {
                   lifecycleManager.getHost(),
                   lifecycleManager.getPort(),
                   celebornConf,
-                  lifecycleManager.getUserIdentifier(),
-                  true);
+                  lifecycleManager.getUserIdentifier());
```
Comment (cfmcgrady) [inline client-spark/spark-3/src/main/java/org/apache/spark/shuffle/celeborn/SparkShuffleManager.java:119]: We don't need to initialize the `shuffleClient` instance at the driver side anymore.

Comment (AngersZhuuuu) [reply client-spark/spark-3/src/main/java/org/apache/spark/shuffle/celeborn/SparkShuffleManager.java:119]: Yea, you are right

**Comment by AngersZhuuuu** [comment]
ping @pan3793 @cfmcgrady @waitinfuture @FMX  Have fix the code after revert, how about current?

File: client-spark/spark-3/src/main/java/org/apache/spark/shuffle/celeborn/SparkShuffleManager.java [hunk 7]
```diff
@@ -151,13 +144,15 @@ public boolean unregisterShuffle(int shuffleId) {
     if (sortShuffleIds.contains(shuffleId)) {
       return sortShuffleManager().unregisterShuffle(shuffleId);
     }
-    if (appUniqueId == null) {
-      return true;
+    // For Spark driver side trigger unregister shuffle.
+    if (lifecycleManager != null) {
+      lifecycleManager.unregisterShuffle(shuffleId);
```
Comment (pan3793) [inline client-spark/spark-3/src/main/java/org/apache/spark/shuffle/celeborn/SparkShuffleManager.java:149]: should we return this value?

File: client-spark/spark-3/src/main/java/org/apache/spark/shuffle/celeborn/SparkShuffleManager.java [hunk 8]
```diff
@@ -151,13 +144,15 @@ public boolean unregisterShuffle(int shuffleId) {
...
     }
-    if (shuffleClient == null) {
-      return false;
+    // For Spark executor side cleanup shuffle related info.
+    if (shuffleClient != null) {
+      shuffleClient.cleanupShuffle(shuffleId);
```
Comment (pan3793) [inline client-spark/spark-3/src/main/java/org/apache/spark/shuffle/celeborn/SparkShuffleManager.java:153]: ditto

File: client/src/main/scala/org/apache/celeborn/client/LifecycleManager.scala [hunk 9]
```diff
@@ -616,7 +611,7 @@ class LifecycleManager(val appUniqueId: String, val conf: CelebornConf) extends
     reply(mapperAttemptFinishedSuccess)
   }
 
-  def handleUnregisterShuffle(
+  def unregisterShuffle(
       shuffleId: Int): Unit = {
```
Comment (pan3793) [inline client/src/main/scala/org/apache/celeborn/client/LifecycleManager.scala:615]: nit: merge to previous line

File: client-spark/spark-3/src/main/java/org/apache/spark/shuffle/celeborn/SparkShuffleManager.java (hunk 7 above)
Comment (AngersZhuuuu) [reply client-spark/spark-3/src/main/java/org/apache/spark/shuffle/celeborn/SparkShuffleManager.java:149]: Not need, in spark's code always return ture

File: client/src/main/scala/org/apache/celeborn/client/LifecycleManager.scala (hunk 9 above)
Comment (AngersZhuuuu) [reply client/src/main/scala/org/apache/celeborn/client/LifecycleManager.scala]: Done

File: client/src/main/scala/org/apache/celeborn/client/LifecycleManager.scala [hunk 10]
```diff
@@ -616,7 +611,7 @@ class LifecycleManager(val appUniqueId: String, val conf: CelebornConf) extends
...
     if (getPartitionType(shuffleId) == PartitionType.REDUCE) {
```
Comment (pan3793) [inline client/src/main/scala/org/apache/celeborn/client/LifecycleManager.scala:615]: @RexXiong curiously, is unregister shuffle required for Flink? or it's not related to engine, just related to MAP/REDUCE partition types

File: client/src/main/java/org/apache/celeborn/client/ShuffleClientImpl.java [hunk 11]
```diff
@@ -192,7 +191,6 @@ public ShuffleClientImpl(String appUniqueId, CelebornConf conf, UserIdentifier u
         ThreadUtils.newDaemonCachedThreadPool(
             "celeborn-shuffle-split", pushSplitPartitionThreads, 60);
     reviveManager = new ReviveManager(this, conf);
-
```
Comment (pan3793) [inline client/src/main/java/org/apache/celeborn/client/ShuffleClientImpl.java:195]: nit: unnecessary change

**Review by pan3793 — APPROVED** [review]

File: client/src/main/java/org/apache/celeborn/client/ShuffleClientImpl.java (hunk 11 above)
Comment (AngersZhuuuu) [reply client/src/main/java/org/apache/celeborn/client/ShuffleClientImpl.java]: DOne

File: client-spark/spark-2/src/main/java/org/apache/spark/shuffle/celeborn/SparkShuffleManager.java [hunk 12]
```diff
@@ -143,13 +136,16 @@ public boolean unregisterShuffle(int shuffleId) {
     if (sortShuffleIds.contains(shuffleId)) {
       return sortShuffleManager().unregisterShuffle(shuffleId);
     }
-    if (appUniqueId == null) {
+    // For Spark driver side trigger unregister shuffle.
+    if (lifecycleManager != null) {
+      lifecycleManager.unregisterShuffle(shuffleId);
       return true;
```
Comment (cfmcgrady) [inline client-spark/spark-2/src/main/java/org/apache/spark/shuffle/celeborn/SparkShuffleManager.java:142]: nit: make it more consistent with spark 3.x shuffle manager

```suggestion
```

**Review by cfmcgrady — APPROVED** [review]

**Comment by pan3793** [comment]
Thanks, merged to main/0.3(for 0.3.1)

File: client-spark/spark-3/src/main/java/org/apache/spark/shuffle/celeborn/SparkShuffleManager.java [hunk 13]
```diff
@@ -189,7 +180,7 @@ public <K, V> ShuffleWriter<K, V> getWriter(
       if (handle instanceof CelebornShuffleHandle) {
         @SuppressWarnings("unchecked")
         CelebornShuffleHandle<K, V, ?> h = ((CelebornShuffleHandle<K, V, ?>) handle);
-        ShuffleClient client =
+        shuffleClient =
```
Comment (waitinfuture) [inline client-spark/spark-3/src/main/java/org/apache/spark/shuffle/celeborn/SparkShuffleManager.java:183]: Should also change this for spark2?
